department-of-veterans-affairs/vets-website
# TOU: Redirect encoding bug + logout

**Note**: Delete the description statements, complete each step. **None are optional**, but can be justified as to why they cannot be completed as written. Provide known gaps to testing that may raise the risk of merging to production.

## Summary
This PR includes the body of work that does the following:
- Removes the `encodeURIComponent` on the window.location change since the URL is already pre-checked in the `parseRedirectUrl`
- Moves the long if/else into the `helpers.js` file as `declineAndLogout`
- Creates the `declineAndLogout` to check if the session storage for `vamobile`
- Creates unit tests for the `declineAndLogout` funciton

## Related issue(s)

- department-of-veterans-affairs/va.gov-team#71408

## Testing done
- Unit + manual testing

## Screenshots
n/a

## What areas of the site does it impact?
This will affect staging & below instances of Terms of Use page, including the VA Mobile app (staging)

## Acceptance criteria

### Quality Assurance & Testing


### Error Handling


### Authentication


### :warning: Team Sites (only applies to modifications made to the VA.gov header) :warning:


## Requested Feedback
**How to test the logout functionality**
1. Sign in with the Sign in Service (`?oauth=true`)
2. Manually navigate to the terms of use page `http://localhost:3001/terms-of-use/?redirect_url=http://localhost:3001/auth/login/callback/?type=idme&terms_code=abc`
3. Change/add in session storage `ci` to `vamobile`
4. Click the Decline and Signout button
5. Confirm the following
    1. You are redirected to the `/terms-of-use/declined` page
    2. Clicking the Sign in button does not open the Sign in modal
    3. Clicking the Sign in button an error in the console that reads: `Failed to launch 'vamobile://login-terms-rejected' because the scheme does not have a registered handler.`


**How to test the redirect bug**
1. Sign in with a user that requires Terms of Use (either SSOe or SiS)
2. After authenticating with a CSP you'll be redirected to the Terms of Use page
3. Confirm the Terms of Use page has the `redirect_url` query parameter pointing towards `/auth/login/callback` page
4. Confirm the `authReturnUrl` session storage item points to the `/auth/login/callback` page
5. Click Accept button > Confirm you are a signed in
